---
title: "Exploration of Phylogenetic RelationshipsBetween Obscure Snake Families"
author: "Nikki Smith/Zizhen Zhong"
date: "2024-10-22/2024-11-15"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

## Introduction

Serpentes is the suborder in Squamata (Lizards) that contains all snakes. Due to their unique morphology, they are an interesting group to study in evolutionary biology. Blind snakes are even more intriguing because of their limited fossil record, strange features and, most importantly, lack of consensus for the branching order of their families (Facini, 2020). Because of these factors, the placement of blind snakes in relation to other snake groups remains an open area of research in phylogenetics.

Previous research has not come to a consensus on the placement of three blind snake families, Anomalepididae (Primitive Blind Snake), Leptotyphlopidae (Threadsnake), Typhlopidae (Typhloids). However, there are some explored possibilities. Depending on the tree used, primative blind snakes are a sister group to all nonblind snakes or a sister group to threadsnakes and typhloids (Singhal, 2021). I wish to explore these relationships further by comparing family placements in a Maximum Likelihood versus a Neighbor-Joining tree as well as how their diets differ. Therefore, I hypothesize that both of the trees will be similar and there will not be a difference in diet between the blind snake families.

```{r Setup, echo= TRUE,results='hide', message=FALSE, warning=FALSE, tidy.opts=list(width.cutoff=60), tidy=TRUE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, results = "hide", message = FALSE, error = FALSE, warning = FALSE)
opts_chunk$set(tidy.opts=list(width.cutoff=60), tidy=TRUE)
# Packages used
library(tidyverse)
library(stringi)
library(ape)
library(RSQLite)
library(Biostrings)
library(muscle)
library(DECIPHER)
library(viridis)

library(rentrez)
library(ggtree)
library(ggplot2)
library(dendextend)
library(styler)
library(phangorn)
library(proxy)

# Squamatabase is the trait database. It was downloaded from Github:
# https://github.com/blueraleigh/squamatabase
# See the accompanying paper too!
# Quite simply, it is a database containing diet samples from the worlds' snakes.
# Each record describes an event of a snake eating or attempting to eat a prey item. 
# It contains various columns with information on the predator and prey, but the relevant ones are "predator_taxon" and "prey". I used this database to match the family to each sequence in 
# sequencedf and also to compare what each relevant snake family eats.
# It is very important to have a personal access token for github or else this will not install.
devtools::install_github("blueraleigh/squamatabase")
library(squamatabase)
```

```{r Data prep, results='hide', message=FALSE, warning=FALSE, tidy.opts=list(width.cutoff=60), tidy=TRUE}
dfDiet <- data(diet)
# Filter out any non-relevant snakes and make a new "family" column
dfFiltered <- squamatabase::filter_records(diet, predator_taxon = "Anomalepididae|Leptotyphlopidae|Typhlopidae")

dfFiltered <- dfFiltered %>%
  mutate(family = sapply(str_split(predator_taxon, ";"), function(x) x[3]))

# Any values in dfFiltered$family containing "rena" are actually "Leptotyphlopidae"
dfFiltered$family[dfFiltered$family == "Rena"] <- "Leptotyphlopidae"

unique(dfFiltered$family)
#==> Notice there are still names in family column not changed 
dfFiltered$family[dfFiltered$family=="Megatyphlops"] <- "Typhlopidae"
dfFiltered$family[dfFiltered$family=="Leptotyphlops"] <- "Leptotyphlopidae"


# Sequences were picked based on 16S ribosomal gene. This is because of the lack of sequence data for Liotyphlops beu, our representitive species for Anomalepididae. This species was chosen since it was the only member of Anomalepididae contained within squamatabase.
relative_path <- "data/serpentes_sequences.fasta"
file_path <- file.path(getwd(),relative_path)
fastaFile <- readDNAStringSet(file_path)

seq_header <- names(fastaFile)
sequence <- paste(fastaFile)
sequencedf <- data.frame(seq_header, sequence)
dim(sequencedf)
#There are 450 samples 

# I extracted the organism name here and made it it's own column.
sequencedf <- sequencedf %>%
  mutate(name = sapply(str_split(seq_header, " "), function(x) {
    paste(x[2], x[3], sep = " ")
  }))

#Noticing names does not include family names
sequencedf$name

# Here we are adding the family names to each sequence in sequencedf using data from dfFiltered.
sequencedf <- sequencedf %>%
  mutate(family = sapply(name, function(x) {
    matched_row <- dfFiltered %>%
      filter(str_detect(predator_taxon, x)) %>%
      pull(family)
    if (length(matched_row) > 0) {
      return(matched_row[1])
    } else {
      return(NA)
    }
  }))

#Checking the new family column to see how the previous function has performed
unique(sequencedf$family)

#Checking counts for unique family names 
table(sequencedf$family)
family_counts <- sequencedf %>% 
  count(family)
view(family_counts)
#==> Noticing there are 421 NAs, and 3 family names we intended to obtain. However, Anomalepididae only has 1 sample. 

#Remove any NAs, duplicates and inserting family names together with their corresponding species names 
sequencedf <- sequencedf %>% 
  filter(!is.na(family) & !duplicated(name)) %>% 
  mutate(family_name = paste(family, name, sep = " ")) 

#confirming the filtration worked
unique(sequencedf$family)
dim(sequencedf)
#==> This unfortunately ends up being a very small dataset with only 8 unique snake species. This reflects the lack of studies in these snakes.

#Turning sequences into biostrings readable, to another column
class(sequencedf$sequence)
sequencedf$sequenceStringset<- DNAStringSet(sequencedf$sequence)
class(sequencedf$sequenceStringset)

#Showing that there's originally no names for the sequences, here giving the names for each biostring sequences with nicely formatted names 
names(sequencedf$sequence)
names(sequencedf$sequenceStringset) <- sequencedf$family_name
names(sequencedf$sequenceStringset)
```

```{r Alignment and GTR model, echo=TRUE,results='hide', tidy.opts=list(width.cutoff=60), tidy=TRUE}
serpentes.alignment <- DNAStringSet(muscle::muscle(sequencedf$sequenceStringset, maxiters = 2), use.names = TRUE)
BrowseSeqs(serpentes.alignment)
#==> Noticing sequences of the 8 species vary dramatically in length. Maybe can consider performing alignment free methods vs alignment methods to see how the phylogenetic tree differs. 


#TN93 model is chosen for generating the distance matrix because it accounts for nucleotide base frequency and transversion/transition bias.
chosen.model <- "TN93"
clustering.threshold <- 0.03

#Generating distance matrix based on the previous Biostring alignment
dnaBin.16S <- as.DNAbin(serpentes.alignment)
distanceMatrix <- dist.dna(dnaBin.16S, model = chosen.model, as.matrix = TRUE, pairwise.deletion = TRUE)
view(distanceMatrix)
#==> Notice samples with same family name generally have < 0.25 distance. Also, distance for the same sample names showed 0, suggesting that this distance matrix has been done correctly. 


# This tree takes about 15 minutes to run on my laptop. GTR stands for general time reversal model. Here I have included the clustering thresholds for generating the first tree. GTR allows for unequal base frequence and assumes time reversibility.  
ml.16S <- DECIPHER::TreeLine(serpentes.alignment, method = "ML", cutoff = clustering.threshold, model = "GTR")

#Plotted the GTR model phylogenetic tree of the 8 species
GTR_model <- as.phylo(ml.16S)

#Here to extract only the family names from that family column, and assign corresponding colours by family name, because family_name column was used previously and assigned to the sequences for alignment
#sequencedf$family_name <- sub(" .*", "", sequencedf$family)
family_tree_colours <- c("Typhlopidae" = "Orange", "Leptotyphlopidae" = "Blue", "Anomalepididae" = "Tomato")
tip_colour <- family_tree_colours[sequencedf$family]

#Plot the GTR phylogenetic tree
GTR_Tree <- plot.phylo(GTR_model, main="Blind Snake Families: 16S rRNA Analysis Using Maximum Likelihood and the GTR Model",cex = 0.85, cex.main= 0.9, tip.color = tip_colour)
#==> Notice the tree has successfully generated 3 major clades, the species are grouped nicely, Anomalepididae family seems to be closely related to both Typhlopidae and Leptotyphylopidae. 
```

```{r NJ model, echo=TRUE,results='hide',fig.keep='all', tidy.opts=list(width.cutoff=60), tidy=TRUE}
nj.16S <- DECIPHER::TreeLine(
  myDistMatrix = distanceMatrix,
  method = "NJ",
  cutoff = clustering.threshold,
  showPlot = FALSE,
  verbose = TRUE
)

#Used another method to plot the phylogenetic tree for NJ model the same way as above
NJ_model <- as.phylo(nj.16S)
NJ_Tree <- plot.phylo(NJ_model, main="Blind Snake Families: 16S rRNA Analysis Using TN93 and Neighbor Joining Model",cex = 0.85, cex.main= 0.9, tip.color = tip_colour)
#==> Similar result is seen as GTR model above. The 3 families are nicely grouped into 3 different clades. Anomalepididae family seems to be closely related to both Typhlopidae and Leptotyphylopidae.
```

```{r GTR & NJ model phylogenetic tree comparision, fig.cap = "Pylogeny Tree Comparison", echo=TRUE,results='hide',fig.keep='all', tidy.opts=list(width.cutoff=60), tidy=TRUE}

# Compare the two dendrograms using a tanglegram. This helps us quickly visualize the conflict in taxa placement between the two trees. We can clearly see that theres a difference in where Leptotyphlopidae and Typhlopidae were placed.
tanglegram(ml.16S, nj.16S,
  main_left = "GTR Dendrogram",
  main_right = "NJ Dendrogram", common_subtrees_color_branches = TRUE, axes = TRUE, lab.cex = 0.89, type = "r", margin_outer = 1, lwd = 0.6)
#==> The result suggests that the phylogenetic relationships among the species in this dataset are largely consistent between the two methods (Maximum Likelihood with GTR and Neighbor-Joining with TN93), with the exception of the relative positions of Typhlopidae Acutotyphlops subocularis and Typhlopidae Typhlops reticulatus.
```

```{r Alignment-Free Method, echo=TRUE,results='hide',tidy.opts=list(width.cutoff=60), tidy=TRUE}
#Here I want to compare the result from alignment free method wit the result seen by alignment methods above, because the sequence length vary dramatically, alignment free method might be more reliable in this case. 

#I want to work with k-mers of 5 in this case 
oligo5 <- oligonucleotideFrequency(x = sequencedf$sequenceStringset, width = 5, as.prob = TRUE)

#Used Euclidean method to calculate distance for alignment free method because it is the most commonly use method in bioinformatic, most straightforward method measuring distance between 2 points in "straight line".  
distance_Alignment_free <- proxy::dist(oligo5, method ="Euclidean")

#To check the distance matrix has been successful made 
print(distance_Alignment_free)

#Using neighbor joining model again, because of its efficieny compared to GTR model. 
NJ_model_free <- nj(distance_Alignment_free)

#Noticing there is no name assigned to the corresponding tip
NJ_model_free$tip.label

#Assigned corresponding names and checked the performance was successful 
NJ_model_free$tip.label <- names(sequencedf$sequenceStringset)
NJ_model_free$tip.label

plot(NJ_model_free, main = "Neighbor Joining Tree Based on 5-mer Frequencies", cex = 0.85, cex.main= 0.9, tip.color = tip_colour)
#==> The results from the alignment-free method are relatively consistent with those from the traditional alignment methods, as samples from the same family are positioned closer to each other. This supports the accuracy of the blind snake species' placements by family names. However, there are some dissimilarity when it comes to the genus names. Notice the Rena genus are not positioned under the same clade. Similarly, Typhlopidae family are also not placed in the same clade, but rather genus Acutotyphlops is shown more closely related to Anomalepididae family. Thus, the results of this study confirms the family-level placement of the blind snakes. However, further research is needed to validate the genus classifications.


```
```{r Blind + Non-Blind Snake Phylogenetic Tree, echo=TRUE,results='hide',tidy.opts=list(width.cutoff=60), tidy=TRUE}

#----Incorporating Non-blind snakes + blind snakes together ---

#This filtering incorporates three additional non-blind snake families for phylogenetic testing. The hypothesis being tested is that Anomalepididae are the sister group to non-blind snakes, as proposed in https://academic.oup.com/sysbio/article/70/3/542/5873536. 

dfFiltered_NB <- squamatabase::filter_records(diet, predator_taxon = "Anomalepididae|Leptotyphlopidae|Typhlopidae|Elapidae|Colubridae|Boidae")

#Separate the predator_taxon column and only obtain the third word from each rows 
dfFiltered_NB <- dfFiltered_NB %>% 
  mutate(family=sapply(str_split(predator_taxon,";"), function(x) x[3]))

unique(dfFiltered_NB$family)
#There are about 88 unique family names, which is too many to manually change the names, thus I will leave this part first 

#Grabbed fasta dataset from NCBI, there are only 24 hits obtained 
NB_search <- entrez_search(db = "nuccore", term = "Boidae[ORGN]| Elapidae[ORGN]| Colubridae[ORGN] AND 16S rRNA [Gene]", retmax = 100)

#These ids all seem differ, good
NB_search$ids

ID.NB <- unique(NB_search$ids)
#In order to loop through and get ids corresponding taxonomic names, has to turn the character class to list class 
ID.NB <- as.list(ID.NB)
class(ID.NB)

#creating an empty vector to store the taxonomic names after the loop. Then looping through each ID to grab their corresponding names to check if samples from 3 families of non-blind snakes are obtained.  
organism.names <- vector("character", length(ID.NB))

for (i in seq_along(ID.NB)) {
  check.ID <- entrez_summary(db = "nuccore", id = ID.NB[i])
  organism.names[i] <- check.ID$organism
}

#There were 33 ids obtained from NCBI, and after matching their corresponding taxonomic names, there are only 13 unique specie names. 
unique(organism.names)

#NB_fetch <- entrez_fetch(db = "nuccore", id = ID.NB, rettype = "fasta")
#write(NB_fetch, "non_blind_snake.fasta", sep = "\n\n")

#Import fasta file and created a new dataframe that include bind and non-blind snake sample sequences 
relative_path <- "data/non_blind_snake.fasta"
file_path <- file.path(getwd(),relative_path)
fastaFile2 <- readDNAStringSet(file_path)

seq_header <- names(fastaFile2)
sequence <- paste(fastaFile2)
sequencedf2 <- data.frame(seq_header, sequence)
View(sequencedf2)

#Creating another column with only sample names info 
sequencedf2 <- sequencedf2 %>%
  mutate(name = sapply(str_split(seq_header, " "), function(x) {
    paste(x[2], x[3], sep = " ")
  }))

#To match the family names 
sequencedf2 <- sequencedf2 %>%
  mutate(family = sapply(name, function(x) {
    matched_row <- dfFiltered_NB %>%
      filter(str_detect(predator_taxon, x)) %>%
      pull(family)
    if (length(matched_row) > 0) {
      return(matched_row[1])
    } else {
      return(NA)
    }
  }))

#By checking the data frame noticed there are 2 species name without corresponding family names and 1 specie had the wrong family name.  Next is to manually include corresponding family names
view(sequencedf2)
unique(sequencedf2$name)

sequencedf2$family[sequencedf2$name == "Sanzinia madagascariensis"] <- "Boidae"
sequencedf2$family[sequencedf2$name == "Hierophis gemonensis" ] <- "Colubridae" 
sequencedf2$family[sequencedf2$name == "Hierophis viridiflavus" ] <- "Colubridae" 
sequencedf2$family[sequencedf2$name == "Pantherophis guttatus" ] <- "Colubridae" 
sequencedf2$family[sequencedf2$name == "Boa constrictor" ] <- "Boidae" 

#No more NAs 
unique(sequencedf2$family)


#In order to merge the 2 dataframes of blind snakes and non-blind snakes, they need same number of columns, merged the family names with specie name
sequencedf2 <- sequencedf2 %>% 
  mutate(family_name = paste(family, name, sep = " "))

#The names have successfully merged
dim(sequencedf2)
unique(sequencedf2$family_name)

sequencedf<- sequencedf[,-6]
combined_sequencedf <- rbind(sequencedf, sequencedf2)
dim(combined_sequencedf)  # 108 rows representing successful merge

#randomly select 2 representative sample per snake family, while allowing families with fewer than 2 samples to include all their available samples. 
ranselect_combined_sequencedf <- combined_sequencedf %>% 
  filter(!duplicated(family_name)) %>% 
  group_by(family) %>% 
  slice_sample(n=2, replace = FALSE) %>% 
  ungroup()

#There are 6 unique family names selected
unique(ranselect_combined_sequencedf$family)

#Assigning names to each sequences
names(ranselect_combined_sequencedf$sequence)
names(ranselect_combined_sequencedf$sequence) <- ranselect_combined_sequencedf$family_name
names(ranselect_combined_sequencedf$sequence) # sucessfull 

#Alignment 
NB.alignment <- DNAStringSet(muscle::muscle(DNAStringSet(ranselect_combined_sequencedf$sequence), maxiters = 2), use.names = TRUE)
BrowseSeqs(NB.alignment)
#==> Noticed a lot of Gap introduced, since there are only 32 samples, cannot perform further filtration to control length. Noticed some names have to incorporate family names. 

#Generating distance matrix based on the previous Biostring alignment
NBdnaBin.16S <- as.DNAbin(NB.alignment)
NB_distanceMatrix <- dist.dna(NBdnaBin.16S, model = chosen.model, as.matrix = TRUE, pairwise.deletion = TRUE)
view(NB_distanceMatrix)

#Generated phylogenetic relationship between non-blind and blind snakes using neighbor joining model, with 0.1 as cluster threshold based on distance matrix. 
NB_nj.16S <- DECIPHER::TreeLine(
  myDistMatrix = NB_distanceMatrix,
  method = "NJ",
  cutoff = 0.03,
  showPlot = FALSE,
  verbose = TRUE
)

#Used another method to plot the phylogenetic tree for NJ model the same way as above
NB_NJ_model <- as.phylo(NB_nj.16S)


family_tree_colours <- c("Typhlopidae" = "Orange", "Leptotyphlopidae" = "Blue", "Anomalepididae" = "Tomato", "Boidae" = "Pink", "Elapidae"= "DarkGreen", "Colubridae" = "Darkgrey")

tip_colour <- family_tree_colours[ranselect_combined_sequencedf$family]

NB_NJ_Tree <- plot.phylo(NB_NJ_model, main="Phylogenetic Relationship Between Blind and Non-blind snakes", tip.color= tip_colour, cex = 0.68, cex.main= 0.9)
#==> I have plotted multiple alignments for the six selected snake families, and the resulting phylogenetic trees vary slightly depending on the species sampled from each family. However, in most of the trees, blind snakes and non-blind snakes form distinct clusters, with related families grouped within the same clade. The placement of Anomalepidae, however, shows some variation. In most cases, the trees do not support the hypothesis that Anomalepidae is a sister group to both blind and non-blind snakes.

```

```{r Prey Consuming Investinagtion, fig.cap = "Prey eaten by snake family. Note the diverse prey sources of Leptotyphlopidae.",echo=TRUE, results= 'hide', tidy.opts=list(width.cutoff=60), tidy=TRUE}
# Now let's visualize what kind of prey members of each of the three families eat. Filter each record by prey and family and get a count of the amount of each prey eaten. I further sorted the data to include prey eaten more than once so I have much less data points. If I chose summary_data$Count > 2, then it would completely eliminate every Anomalepididae record, which is not what we want.
summary_data <- dfFiltered %>%
  group_by(family, prey) %>%
  summarize(Count = n())

#Began with 50 samples 
dim(summary_data)

#Obtain only samples with more than 1 prey counts
summary_data <- summary_data[summary_data$Count > 1, ]
dim(summary_data)  #22 samples remaining 

# From the plot we can see that Formicidae (ants) are the most common prey across families. Leptotyphlopidae has the most diverse diet among the three groups, while Anomalepididae is the least diverse. There does not appear to be a correlation between diet and phylogenetic placement.
ggplot(summary_data, aes(x = prey, y = Count, fill = family)) +
  geom_bar(stat = "identity", position = "dodge") +
  scale_fill_viridis_d() +
  labs(
    title = "Distribution of Prey Items by Predator Family",
    x = "Prey",
    y = "Frequency", fill = "Predator Family"
  )+
  theme(axis.text.x = element_text(angle = 45, hjust = 1), plot.title = element_text(hjust = 0.5))
```

## Discussion & Conclusion

Based on Figure 1 and Figure 2, we can see a distinct difference in the placement of primitive blind snakes. In the ML tree, primitive blind snakes are a sister group only to threadsnakes, while in the NJ tree, primitive blind snakes are a sister group only to typhloids. This is a contrast to what was found in previous literature, where the ML tree found that primitive blind snakes were a sister group to all non-blind snakes (Singhal, 2021). I also found that threadsnakes had a more diverse diet than the other two families, but otherwise, diet differences did not reflect in their phylogenetic placements. Therefore, I reject my original hypothesis.

There were several limitations to my analysis. The most striking one is the very small sequence sample size (n = 8). This makes it more difficult to make a descriptive tree. It also represents bias in the NCBI database, since it likely contains more records for snakes with medical-significance (i.e venomous snakes). Also, since Liotyphlops beu was the only primitive blind snake in squamatabase, it narrowed down the genes I could use for alignment. This species only had two nucleotide sequences available, so I had to use 16S ribosomal RNA gene. Another limitation is the lack of diet records in squamatabase. For primitive blind snakes in particular, there were only 3 records out of a total of 11049 observations. This makes diet analysis a lot harder to do. Future studies would benefit from having a more robust array of species and sequences. Also, the use of neural networks could be promising in the future, similar to what was done by Burbrink et al. (2020).

## Acknowledgements

Code credits:

<https://carpentries.github.io/sandpaper-docs/github-pat.html> <https://stackoverflow.com/questions/21263636/read-fasta-into-a-dataframe-and-extract-subsequences-of-fasta-file> <https://stackoverflow.com/questions/5559467/how-to-merge-two-columns-in-r-with-a-specific-symbol> <https://stackoverflow.com/questions/64160683/how-to-add-parentheses-to-column-values-in-a-data-frame-in-r>
<https://rdrr.io/cran/ape/man/plot.phylo.html>
<https://eeob-macroevolution.github.io/Practicals/Intro_to_Phylo/intro_to_phylo.html>

## References

1.  Fachini, T. S., Onary, S., Palci, A., Lee, M. S. Y., Bronzati, M., & Hsiou, A. S. (2020). Cretaceous Blind Snake from Brazil Fills Major Gap in Snake Evolution. iScience, 23(12), 101834. <https://doi.org/10.1016/j.isci.2020.101834>
2.  Singhal, S., Colston, T. J., Grundler, M. R., et al. (2021), Congruence and Conflict in the Higher-Level Phylogenetics of Squamate Reptiles: An Expanded Phylogenomic Perspective, Systematic Biology, Volume 70, Issue 3, May 2021, Pages 542–557, <https://doi.org/10.1093/sysbio/syaa054>
3.  Grundler MC (2020) SquamataBase: a natural history database and R package for comparative biology of snake feeding habits. Biodiversity Data Journal 8: e49943. <https://doi.org/10.3897/BDJ.8.e49943>
4.  Burbrink, F.T., Grazziotin, F. G., Pyron, R A., et al. (2020), Interrogating Genomic-Scale Data for Squamata (Lizards, Snakes, and Amphisbaenians) Shows no Support for Key Traditional Morphological Relationships, Systematic Biology, Volume 69, Issue 3, May 2020, Pages 502–520, <https://doi.org/10.1093/sysbio/syz062>
